### PR TITLE
feat(u): Hexstring class

### PR DESCRIPTION
**Pull request by MorganDream**
This is tryout for solution to https://github.com/CityOfZion/neon-js/issues/480.
I have created below types:
- HexString
- ScriptHash
- PublicKey
- Address
And I have tried to use them in Class Cosigner and TransactionAttribute.

And Let's talk about several design issues about this feature:
1. Where should we put these types, and what's the relation between these types and old type-converting apis.
Suggestion: We put these types into module `u`, and stop exporting old type-conversion apis like `str2hextring` and include them into these types, to avoid confusions.

2. `-Like` interfaces
The `-Like` interfaces (e.g. CosignerLike) mostly use string as type, and they will be used to type param for constructors. So shall change the `-Like` interfaces to use new types.
Currently I create a new interface `-Config` to act as the param for constructor.

### CODE REVIEW

**Comment by snowypowers** [comment]
Please focus on the hexstring class first. The others can be filled in later.

**Comment by MorganDream** [comment]
Now only HexString class is left. And let's settle this first.

File: packages/neon-core/src/u/HexString.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    this._value = value;
+    this._littleEndian = littleEndian;
+  }
+
+  /**
+   * Get hex value as string
+   * @param asLittleEndian true, export a little-endian string; false, export a big-endian string; default to be false
+   * @returns string
+   */
+  public value(asLittleEndian = false): string {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: replace `value` and `reverse` with `toLittleEndian` and `toBigEndian`. Alternatively, we can use `toString` and `toReverseString` as the API instead.

File: packages/neon-core/src/u/HexString.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  public toNumber(): number {
+    return parseInt(this.value(), 16);
+  }
+
+  public toArrayBuffer(): Uint8Array {
+    return hexstring2ab(this.value());
+  }
+
+  public static fromHexString(str: string, littleEndian = false) {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: fromHex

File: packages/neon-core/src/u/HexString.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    return new HexString(str, littleEndian);
+  }
+
+  public static fromASCIIString(str: string, littleEndian = false) {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: fromAscii(). And remove the littleEndian argument.

File: packages/neon-core/src/u/HexString.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    const hex = str2hexstring(str);
+    return new HexString(littleEndian ? reverseHex(hex) : hex, littleEndian);
+  }
+
+  public static fromNumber(num: number, littleEndian = false) {
+    return new HexString(num2hexstring(num, 1, littleEndian), littleEndian);
+  }
+
+  public static fromArrayBuffer(arr: ArrayBuffer | ArrayLike<number>) {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: I think arraybuffers also have an endianness.

File: packages/neon-core/src/u/HexString.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    ensureHex(value);
+  }
+
+  protected constructor(value: string, littleEndian = false) {
+    if (value.startsWith("0x")) {
+      value = value.slice(2);
+    }
+    this._checkValue(value);
+    this._value = value;
+    this._littleEndian = littleEndian;
```
[the remark this answers is not in the document]
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: Yeah, good idea.

File: packages/neon-core/src/u/HexString.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+  public isEqualTo(other: HexString): boolean {
+    return this.value() === other.value();
+  }
+
+  public xor(other: HexString): HexString {
+    return HexString.fromHexString(hexXor(this.value(), other.value()));
+  }
+
+  public toASCIIString(): string {
```
[the remark this answers is not in the document]
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: Changed

File: packages/neon-core/src/u/HexString.ts (hunk 1 above)
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: The `value` can be replaced with `toLittleEndian` and `toBigEndian` but not `reverse`.
`reverse` is a write function to change the value of HexString.

I used `export(asLittleEndian: boolean)` to replace `value` cause I think `toLittleEndian` and `toBigEndian` sounds like they're going to do some conversion to change the value.

File: packages/neon-core/src/u/HexString.ts (hunk 3 above)
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: I also removed littleEndian param for `fromNum`

File: packages/neon-core/src/u/HexString.ts (hunk 4 above)
Comment (snowypowers) [reply packages/neon-core/src/u/HexString.ts]: the user will have to specify the endianness on creation. Its similar to `fromHex`.

File: packages/neon-core/src/u/HexString.ts (hunk 1 above)
Comment (snowypowers) [reply packages/neon-core/src/u/HexString.ts]: Remove the write function and we have an immutable class that is very clear in terms of API. If we need to reverse the internal value, it means that we stored the value wrongly. There should never be a need to reverse the value.

We should always be stating how we are using the hexstring values so `toLittleEndian` and `toBigEndian` works nicely.

File: packages/neon-core/src/u/HexString.ts (hunk 3 above)
Comment (snowypowers) [reply packages/neon-core/src/u/HexString.ts]: this is slightly debatable as a number representation can be in either endian.

File: packages/neon-core/src/u/HexString.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+  public toBigEndian(): string {
+    return this._value;
+  }
+
+  public toLittleEndian(): string {
+    return reverseHex(this._value);
+  }
+
+  public reverse(): this {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: please remove this. There should be no cases where this comes in useful.

File: packages/neon-core/src/u/HexString.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    this._value = reverseHex(this._value);
+    return this;
+  }
+
+  public isEqualTo(other: HexString): boolean {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: we use `equals` on all our other classes so stick with the convention.

File: packages/neon-core/src/u/HexString.ts (hunk 7 above)
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: OK

File: packages/neon-core/src/u/HexString.ts (hunk 8 above)
Comment (MorganDream) [reply packages/neon-core/src/u/HexString.ts]: OK

File: packages/neon-core/src/u/HexString.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+   * Export as ASCII string
+   */
+  public toAscii(): string {
+    return hexstring2str(this.toBigEndian());
+  }
+
+  /**
+   * Export as number
+   */
+  public toNumber(): number {
```
Comment (snowypowers) [inline packages/neon-core/src/u/HexString.ts]: if we have the option to import with an custom endian option, we should allow export with the same custom option. Similar for arraybuffers.

File: packages/neon-core/__tests__/u/HexString.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    const hex = HexString.fromHex("a4c2");
+    const arr = hex.toArrayBuffer();
+    expect(arr[0]).toBe(164);
+    expect(arr[1]).toBe(194);
+  });
+});
+
+describe("math operation", () => {
+  test("xor", () => {
+    const hex = HexString.fromHex("a4c2");
```
Comment (snowypowers) [inline packages/neon-core/__tests__/u/HexString.ts]: pls add an comment describing the bits so its easier to see the effects

```
const hex = HexString.fromHex("a4c2"); //1010 0100 1100 0010
```

Comment (MorganDream) [reply packages/neon-core/__tests__/u/HexString.ts]: All modified.

**Comment by hal0x2328** [comment]
300 points morgandream
200 points snowypowers for review

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/44C84C9E7582DC00411BC9D9D075F6993BA7E203FFDDE366E21DFEBD125EA580

**Pull request merged**